src/ck-ldb/lb_strategy.h

```diff
@@ -0,0 +1,235 @@
+#ifndef LB_STRATEGY_H
+#define LB_STRATEGY_H
+
+#include <random>
+#include <algorithm>
+
+namespace lb_strategy
+{
+
+  template <typename T, bool is_ptr = std::is_pointer<T>::value> struct CmpLoadGreater {};
+
+  template <typename T> struct CmpLoadGreater<T,true> {
+    inline bool operator()(const T a, const T b) const { return (a->getLoad() > b->getLoad()); }
+  };
+
+  template <typename T> struct CmpLoadGreater<T,false> {
+    inline bool operator()(const T &a, const T &b) const { return (a.getLoad() > b.getLoad()); }
+  };
+
+  template <typename T, bool is_ptr = std::is_pointer<T>::value> struct CmpLoadLess {};
+
+  template <typename T> struct CmpLoadLess<T,true> {
+    inline bool operator()(const T a, const T b) const { return (a->getLoad() < b->getLoad()); }
+  };
+
+  template <typename T> struct CmpLoadLess<T,false> {
+    inline bool operator()(const T &a, const T &b) const { return (a.getLoad() < b.getLoad()); }
+  };
+
+  template <typename T, bool is_ptr = std::is_pointer<T>::value> struct CmpId {};
+
+  template <typename T> struct CmpId<T,true> {
+    inline bool operator()(const T a, const T b) const { return (a->id < b->id); }
+  };
+
+  template <typename T> struct CmpId<T,false> {
+    inline bool operator()(const T &a, const T &b) const { return (a.id < b.id); }
+  };
+
+  template <typename T>
+  T *ptr(T &obj) { return &obj; }  // turn reference into pointer
+
+  template <typename T>
+  T *ptr(T *obj) { return obj; }  // obj is already pointer, return it
+
+  // ---------------- Obj --------------------
+
+  struct obj_1_data { float load; };
+  template <int N> struct obj_N_data { float load[N]; float maxload; };
+
+  template <int N, bool multi = (N > 1)>
+  class Obj : public std::conditional<multi, obj_N_data<N>, obj_1_data>::type
+  {
+  public:
+
+    unsigned int id;
```

Comment:
`unsigned` is often not a good idea. Is there a reason why `id` is unsigned here (in contrast to Proc's id, for example)?

Reply:
For objects, this ID is basically the index in the vector of objects passed to lb strategies. It is not related to 64 bit ID of chares. So, it is unsigned because it goes from 0 to numobjs-1. But I agree that `unsigned int` is not ideal. It should probably need to be typedef'd to some unsigned type though. The size of the data type will determine how many objects a load balancing strategy can accept. I'm not sure if we need a 64-bit type. It could make running a strategy like greedy with millions of objects slightly slower compared to 32-bit. Also, very big jobs would presumably be using a hierarchical mode of TreeLB so that top-level strategies would not deal with many millions of objects.

Reply:
Proc has a signed ID because it coincides with the PE number, which is `int`, and also if I remember correctly the PE number of some objects can be unknown if the migrate from one subtree to another.

Reply:
I've updated this to use regular `int`, since that's what the type of the incoming value that set this is.